File: scala/powercards_replay/Powercards.scala
package core

import collection.immutable.Queue

object Powercards {
  
  class RestoreContext(
    var checkpoint: Game,
    var pendingUserInputs: Queue[Vector[Int]],
    var processedUserInputs: Queue[Vector[Int]]
  ) {
    def ask[A](player: Player, message: String, choices: Vector[A], selectable: A => Boolean)
      : (Vector[A], Vector[A]) =
      if (!this.pendingUserInputs.isEmpty) {
        val (input, _) = this.pendingUserInputs.dequeue
        val inputset = input.toSet
        val (selected, nonselected) = choices.zipWithIndex.partition { x => inputset(x._2) }
        (selected.map { _._1 }, nonselected.map { _._1 })
      } else
        throw pause(player, message, choices)
    
    def pause[A](player: Player, message: String, choices: Vector[A]): AskException = 
      new AskException(new AskDialog(player, message, choices.map(_.toString)))
    
    def approve() {
      val (input, otherInputs) = this.pendingUserInputs.dequeue
      this.processedUserInputs = this.processedUserInputs.enqueue(input)
      this.pendingUserInputs = otherInputs
    }
    
    def markCheckpoint(game: Game) {
      this.checkpoint = new Game(
          players = game.players.map(p => new Player(
            name = p.name, 
            stat = new PlayerStat(actions = p.stat.actions, buys = p.stat.buys, coins = p.stat.coins),
            stage = p.stage,
            deck = p.deck,
            hand = p.hand,
            played = p.played,
            discard = p.discard
          )), 
          activePlayerIndex = game.activePlayerIndex,
          trash = game.trash
        )
      this.processedUserInputs = Queue.empty
    }
  } 

  class Game(
    val players: Vector[Player],
    var activePlayerIndex: Int,
    var trash: Vector[Card]
  ) {
    def activePlayer: Player = this.players(this.activePlayerIndex)
    
    def progress(userInputs: Queue[Vector[Int]]): (AskDialog, RestoreContext) = {
      
      @annotation.tailrec
      def loop(context: RestoreContext) {
        context.markCheckpoint(this)
        this.activePlayer.stage match {
          case PlayerStage.Action => 
            val actionCards = this.activePlayer.hand collect { case card: Actionable => card }
            if (!actionCards.isEmpty) {
              val cards = context.ask(this.activePlayer, "choose an action card to play", actionCards)
              if (cards.isEmpty) {
                context.approve()
                this.activePlayer.stage = PlayerStage.Treasure
                loop(context)
              } else if (cards.length == 1) {
                context.approve()
                val actionCard = cards(0)
                moveCard(actionCard, this.activePlayer.hand, this.activePlayer.played) { (from, to) => 
                  this.activePlayer.hand = from
                  this.activePlayer.played = to
                }
                actionCard.play(this, context)
                loop(context)
              } else
                throw context.pause(this.activePlayer, "you can only choose one action card to play", actionCards)
            } else {
              this.activePlayer.stage = PlayerStage.Treasure
              loop(context)
            }
          case PlayerStage.Treasure => 
            val treasureCards = this.activePlayer.hand collect { case card: Treasurable => card }
            if (!treasureCards.isEmpty) {
              val cards = context.ask(this.activePlayer, "choose treasure cards to play", treasureCards)
              if (cards.isEmpty) {
                context.approve()
                this.activePlayer.stage = PlayerStage.Buy
                loop(context)
              } else {
                context.approve()
                moveCards(cards, this.activePlayer.hand, this.activePlayer.played) { (from, to) => 
                  this.activePlayer.hand = from
                  this.activePlayer.played = to
                }
                this.activePlayer.stat.coins += cards.map(_.coins).sum
                loop(context)
              }
            } else {
              this.activePlayer.stage = PlayerStage.Buy
              loop(context)
            }
        }
      }
      
      val context = new RestoreContext(
        checkpoint = null,
        pendingUserInputs = userInputs,
        processedUserInputs = Queue.empty
      ) 
      
      try {
        loop(context)
        throw new Exception("game loop is incomplete")
      } catch {
        case ex: AskException => (ex.dialog, context)
      }
    }
    
    def moveCard(card: Card, from: Vector[Card], to: Vector[Card])
      (assignment: (Vector[Card], Vector[Card]) => Unit) {
      val from1 = from.filter { _ != card }
      val to1 = to :+ card
      assignment(from1, to1)
    }
    
    def moveCards(cards: Vector[Card], from: Vector[Card], to: Vector[Card])
      (assignment: (Vector[Card], Vector[Card]) => Unit) {
      val from1 = from.filter { !cards.toSet(_) }
      val to1 = to ++ cards
      assignment(from1, to1)
    }
  }
  
  class Player(
    val name: String,
    val stat: PlayerStat,
    var stage: PlayerStage.Value,
    var deck: Vector[Card],
    var hand: Vector[Card],
    var played: Vector[Card],
    var discard: Vector[Card]
  ) {
    
    def activate() {
      this.stat.actions = 1
      this.stat.buys = 1
      this.stat.coins = 0
      this.stage = PlayerStage.Action
    }
    
    @annotation.tailrec
    final def drawCards(count: Int): Boolean = {
      if (count == 0) true
      else if (this.deck.isEmpty && this.discard.isEmpty) false
      else if (this.deck.isEmpty) {
        this.deck = util.Random.shuffle(this.discard)
        this.discard = Vector.empty
        drawCards(count)
      }
      else {
        this.hand :+= this.deck.last
        this.deck = this.deck.dropRight(1)
        drawCards(count - 1)
      }
    }
    
    override def toString: String = {
      val builder = new collection.mutable.StringBuilder
      builder.append("name %s stat %s stage %s \n".format(this.name, this.stat, this.stage))
      builder.append("deck %s \n".format(this.deck))
      builder.append("hand %s \n".format(this.hand))
      builder.append("played %s \n".format(this.played))
      builder.append("discard %s \n".format(this.discard))
      builder.toString
    }
  }
  
  object Game {
    private val initDeck: Vector[Card] = Vector.fill(3)(Estate) ++ Vector.fill(7)(Copper)
    
    def create(playerNames: Vector[String]): Game = {
      val players = playerNames map Player.create 
      
      players foreach { player =>
        player.deck = util.Random.shuffle(initDeck)
        player.drawCards(5)
      }
      
      val game = new Game(
        players = players,
        activePlayerIndex = util.Random.nextInt(playerNames.length),
        trash = Vector.empty
      )
      game.activePlayer.activate()
      game
    }
  }
  
  object Player {
    def create(name: String): Player = 
      new Player(
        name = name, 
        stat = new PlayerStat(actions = 0, buys = 0, coins = 0),
        stage = PlayerStage.Inactive,
        deck = Vector.empty,
        hand = Vector.empty,
        played = Vector.empty,
        discard = Vector.empty
      )
  }
  
  class PlayerStat(var actions: Int, var buys: Int, var coins: Int) {
    override def toString: String = "actions:%d buys:%d coins:%d".format(this.actions, this.buys, this.coins)
  }
  
  object PlayerStage extends Enumeration {
    val Inactive, Action, Treasure, Buy, Cleanup = Value
  }
  
  class AskDialog(
    val player: Player,
    val message: String,
    val choices: Vector[String]
  )
  
  class AskException(val dialog: AskDialog) extends Exception
  
  trait Card {
    def name: String
    def cost: Int
    
    override def toString: String = name
  }
  
  trait Actionable {
    def play(game: Game, context: RestoreContext)
  }
  
  trait Treasurable {
    def coins: Int
  }
  
  object Copper extends Card with Treasurable {
    val name = "Copper"
    val cost = 0 
    val coins = 1
  }
  
  object Estate extends Card {
    val name = "Estate"
    val cost = 2
  }
  
  object Remodel extends Card with Actionable {
    val name = "Remodel"
    val cost = 4
    
    def play(game: Game, context: RestoreContext) {
      println("playing " + this)
      
      if (game.activePlayer.hand.isEmpty) {
        println("you have no card in hand to trash")
        return
      }
        
      val cardsToTrash = context.ask(game.activePlayer, "choose a card to trash", game.activePlayer.hand)
      if (cardsToTrash.isEmpty)
        throw context.pause(game.activePlayer, "you must choose one card to trash", game.activePlayer.hand)
      
      if (cardsToTrash.length > 1)  
        throw context.pause(game.activePlayer, "you can only choose one card to trash", game.activePlayer.hand)
      
      context.approve()
      val cardToTrash = cardsToTrash(0)
      
      game.moveCard(cardToTrash, game.activePlayer.hand, game.trash) { (from, to) =>
        game.activePlayer.hand = from
        game.trash = to
      }
      println("trashed " + cardToTrash)
      
      val supply = Vector(Remodel, ThroneRoom)
      val cardsToGain = context.ask(game.activePlayer, "gain a card up to %d coins".format(cardToTrash.cost + 2), supply)
      if (cardsToGain.isEmpty)
        throw context.pause(game.activePlayer, "you must gain a card up to %d coins".format(cardToTrash.cost + 2), supply)
      
      if (cardsToGain.length > 1)
        throw context.pause(game.activePlayer, "you can only gain one card up to %d coins".format(cardToTrash.cost + 2), supply)
        
      context.approve()
      val cardToGain = cardsToGain(0)
      
      game.activePlayer.discard :+= cardToGain
      println("gained " + cardToGain)
    }
  }
  
  object ThroneRoom extends Card with Actionable {
    val name = "Throne Room"
    val cost = 4
    
    def play(game: Game, context: RestoreContext) {
      println("playing " + this)
      
      val actionCards = game.activePlayer.hand collect { case card: Actionable => card }
      if (actionCards.isEmpty) {
        println("you have no action card in hand to play")
        return
      }
      
      val cardsToPlay = context.ask(game.activePlayer, "choose an action card to play twice", actionCards)
      if (cardsToPlay.isEmpty)
        throw context.pause(game.activePlayer, "you must choose one card to play twice", actionCards)
        
      if (cardsToPlay.length > 1)
        throw context.pause(game.activePlayer, "you must only choose one card to play twice", actionCards)
        
      context.approve()
      val cardToPlay = cardsToPlay(0)
      game.moveCard(cardToPlay, game.activePlayer.hand, game.activePlayer.played) { (from, to) => 
        game.activePlayer.hand = from
        game.activePlayer.played = to
      }
      
      (1 to 2).foreach { _ => cardToPlay.play(game, context) }
    }
  }
}

File: ocaml/powercards/powercards.ml
open Core.Std

type stage = Action | Treasure | Buy | Cleanup

type io = {
  input: unit -> string;
  output: string -> unit;
}
type requirement = Unlimited | MandatoryOne | OptionalOne
type choice = Unselectable | Skip | Indexes of int list

type stat = {
  mutable actions: int;
  mutable buys: int;
  mutable coins: int;
}

type game = {
  players: player list;
  mutable active_player_index: int;
  piles: pile list;
  mutable trash: card list;
  mutable stage: stage;
  stat: stat;
}
and player = {
  name: string;
  mutable deck: card list;
  mutable hand: card list;
  mutable played: card list;
  mutable discard: card list;
}
and pile = {
  sample: card;
  mutable size: int;
}
and card = {
  title: string;
  cost: int;
  feature: card_feature;
}
and card_feature =
  | BasicAction of (io -> game -> unit)
  | SelfTrashAction of (io -> game * card * bool -> bool)
  | BasicTreasure of int
  | BasicVictory of int
  | DynamicVictory of (player -> int)

let rec loop_til ~f = match f () with
  | Some x -> x
  | None -> loop_til ~f

let all_preds predicates x = List.for_all predicates ~f:(fun p -> p x)

let active_player game =
  List.nth_exn game.players game.active_player_index

let choose io requirement message items =
  let select_one input =
    let index = int_of_string input in
    match List.nth_exn items index with
    | (_, true) -> Some (Indexes [index])
    | (name, false) ->
      io.output (sprintf "%s is not selectable" name);
      None
  in
  let select_many input =
    let indexes = input
                  |> String.split ~on:','
                  |> List.map ~f:(String.strip ?drop:None)
                  |> List.filter ~f:(Fn.non String.is_empty)
                  |> List.map ~f:int_of_string
                  |> List.sort ~cmp:Int.compare
    in
    let nonSelectables = indexes
                         |> List.filter_map ~f:(fun i ->
                             let (title, selectable) = List.nth_exn items i in
                             if selectable then None else Some title)
    in
    if List.is_empty nonSelectables then
      Some (Indexes indexes)
    else begin
      nonSelectables
      |> String.concat ~sep:", "
      |> sprintf "%s are not selectable"
      |> io.output;
      None
    end
  in
  let ask () =
    io.output message;
    List.iteri items ~f:(fun i item ->
        sprintf "[%d] %s %s" i (fst item) (if snd item then "(select)" else "")
        |> io.output);
    match requirement with
    | MandatoryOne -> io.input () |> select_one
    | OptionalOne ->
      io.output "or skip";
      begin match io.input () with
      | "skip" -> Some Skip
      | input -> select_one input
      end
    | Unlimited ->
      io.output "or all, or skip";
      match io.input () with
      | "skip" -> Some Skip
      | "all" ->
        let indexes = List.filter_mapi items ~f:(fun i (_, selectable) ->
            if selectable then Some i else None)
        in Some (Indexes indexes)
      | input -> select_many input
  in
  if List.exists items ~f:snd then loop_til ~f:ask else Unselectable

let copper = { title = "Copper"; cost = 0; feature = BasicTreasure 1 }
let silver = { title = "Silver"; cost = 3; feature = BasicTreasure 2 }
let gold = { title = "Gold"; cost = 6; feature = BasicTreasure 3 }

let estate = { title = "Estate"; cost = 2; feature = BasicVictory 1 }
let duchy = { title = "Duchy"; cost = 5; feature = BasicVictory 3 }
let province = { title = "Province"; cost = 8; feature = BasicVictory 6 }

let create_player name =
  let deck = 
    let estates = List.init 3 ~f:(fun _ -> estate) in
    let coppers = List.init 7 ~f:(fun _ -> copper) in
    List.permute (estates @ coppers)
  in
  { name;
    deck = List.take deck 5;
    hand = List.drop deck 5;
    played = [];
    discard = [];
  }

let create_start_piles kingdom_cards player_cnt =
  let pile size sample = { sample; size } in
  let treasures = [pile (60 - player_cnt * 7) copper; pile 40 silver; pile 30 gold] in
  let victories =
    let n = match player_cnt with
      | 2 -> 8
      | 3 | 4 -> 12
      | _ -> assert false
    in [pile n estate; pile n duchy; pile n province]
  in
  let kingdoms = List.map kingdom_cards ~f:(pile 10) in
  treasures @ victories @ kingdoms

let is_action card = match card.feature with
  | BasicAction _ -> true
  | SelfTrashAction _ -> true
  | BasicTreasure _ -> false
  | BasicVictory _ -> false
  | DynamicVictory _ -> false

let is_treasure card = match card.feature with
  | BasicAction _ -> false
  | SelfTrashAction _ -> false
  | BasicTreasure _ -> true
  | BasicVictory _ -> false
  | DynamicVictory _ -> false

let is_victory card = match card.feature with
  | BasicAction _ -> false
  | SelfTrashAction _ -> false
  | BasicTreasure _ -> false
  | BasicVictory _ -> true
  | DynamicVictory _ -> true

let same_kind x y = x.title = y.title

let find_pile game card = game.piles |> List.find_exn ~f:(fun p -> same_kind p.sample card)

let pile_empty pile = pile.size = 0

let split_by_indexes xs indexes =
  let aux i acc x = match acc with
    | (yes, no, hd :: tl) when i = hd -> (x :: yes, no, tl)
    | (yes, no, l) -> (yes, x :: no, l)
  in
  let (yes, no, _) = List.foldi xs ~f:aux ~init:([], [], indexes) in
  (List.rev yes, List.rev no)

let rec draw_card player = match (player.deck, player.discard) with
  | ([], []) -> None
  | ([], discard) ->
    player.deck <- List.permute discard;
    player.discard <- [];
    draw_card player
  | (hd :: tl, _) ->
    player.deck <- tl;
    player.hand <- hd :: player.hand;
    Some hd

let draw_cards_until player predicate =
  let rec loop acc =
    if predicate acc then List.rev acc
    else match draw_card player with
      | None -> List.rev acc
      | Some card -> loop (card :: acc)
  in
  loop []

let draw_cards player n = draw_cards_until player (fun cards -> List.length cards >= n)

let is_ended game =
  let cond1 () = find_pile game province |> pile_empty in
  let cond2 () = game.piles |> List.count ~f:pile_empty |> (<=) 3 in
  cond1 () || cond2 ()

let end_stat game =
  let card_vps p c = match c.feature with
    | BasicAction _ | SelfTrashAction _ | BasicTreasure _ -> 0
    | BasicVictory x -> x
    | DynamicVictory f -> f p
  in
  let player_vps p =
    [p.deck; p.hand; p.played; p.discard]
    |> List.concat
    |> List.fold ~init:0 ~f:(fun acc c -> acc + (card_vps p c))
  in
  let stats = game.players
              |> List.map ~f:(fun p -> (p.name, player_vps p))
              |> List.sort ~cmp:(fun (_, vps1) (_, vps2) -> Int.compare vps1 vps2)
              |> List.rev
  in
  let max_vps = List.hd_exn stats |> snd in
  stats |> List.map ~f:(fun (name, vps) -> (name, vps, (vps = max_vps)))

let card_to_item predicate card = (card.title, predicate card)

let pile_to_item predicate pile = (pile.sample.title, predicate pile)

let gain_card io game player predicate =
  match game.piles
        |> List.map ~f:(pile_to_item predicate)
        |> choose io MandatoryOne "Select a pile to gain"
  with
  | Unselectable ->
    io.output "No pile available to gain";
  | Indexes [index] ->
    let pile = List.nth_exn game.piles index in
    pile.size <- pile.size - 1;
    player.discard <- pile.sample :: player.discard;
    "Gained " ^ pile.sample.title |> io.output;
  | Indexes _ | Skip -> assert false

let play_one io game =
  let play_cards indexes =
    let player = active_player game in
    let (cards, hand) = split_by_indexes player.hand indexes in
    player.hand <- hand;
    player.played <- cards @ player.played;
    cards
  in
  let buy_card index =
    let player = active_player game in
    let pile = List.nth_exn game.piles index in
    pile.size <- pile.size - 1;
    player.discard <- pile.sample :: player.discard;
    pile.sample
  in
  let play_action () =
    let skip () =
      game.stat.actions <- 0;
      game.stage <- Treasure
    in
    if game.stat.actions > 0 then begin
      match (active_player game).hand
            |> List.map ~f:(card_to_item is_action)
            |> choose io OptionalOne "Select an action card to play"
      with
      | Unselectable ->
        io.output "No action card to play. Skip to treasure stage";
        skip ()
      | Skip ->
        io.output "Skip to treasure stage";
        skip ()
      | Indexes indexes ->
        game.stat.actions <- game.stat.actions - 1;
        match play_cards indexes with
        | [card] ->
          "Playing " ^ card.title |> io.output;
          begin match card.feature with
          | BasicAction play -> play io game
          | SelfTrashAction play -> play io (game, card, false) |> ignore
          | BasicTreasure _ | BasicVictory _ | DynamicVictory _ -> assert false
          end
        | _ -> assert false
    end else begin
      io.output "No action point. Skip to treasure stage";
      skip ()
    end
  in
  let play_treasure () =
    let skip () =
      game.stat.buys <- 0;
      game.stage <- Buy
    in
    begin match (active_player game).hand
          |> List.map ~f:(card_to_item is_treasure)
          |> choose io Unlimited "Select treasure cards to play"
    with
    | Unselectable ->
      io.output "No treasure card to play. Skip to buy stage";
      skip ()
    | Skip ->
      io.output "Skip to buy stage";
      skip ()
    | Indexes indexes ->
      let cards = play_cards indexes in
      cards
      |> List.map ~f:(fun c -> c.title)
      |> String.concat ~sep:","
      |> (^) "playing "
      |> io.output;
      List.iter cards ~f:(fun c -> match c.feature with
          | BasicTreasure coins -> game.stat.coins <- game.stat.coins + coins
          | BasicAction _ | SelfTrashAction _ | BasicVictory _ | DynamicVictory _ ->
            assert false
        )
    end
  in
  let play_buy () =
    let skip () =
      game.stat.coins <- 0;
      game.stage <- Cleanup
    in
    if game.stat.buys > 0 then begin
      match game.piles
            |> List.map ~f:([Fn.non pile_empty; (fun p -> p.sample.cost <= game.stat.coins)] |> all_preds |> pile_to_item)
            |> choose io OptionalOne "Select a pile to buy"
      with
      | Unselectable ->
        io.output "No card to buys. Skip to cleanup stage";
        skip ()
      | Skip ->
        io.output "Skip to cleanup stage";
        skip ()
      | Indexes [index] ->
        game.stat.buys <- game.stat.buys - 1;
        let card = buy_card index in
        game.stat.coins <- game.stat.coins - card.cost;
        "Bought " ^ card.title |> io.output
      | Indexes _ -> assert false
    end else begin
      io.output "No more buys. Skip to cleanup stage";
      skip ()
    end
  in
  let play_cleanup () =
    let player = active_player game in
    player.discard <- player.played @ player.hand @ player.discard;
    player.played <- [];
    player.hand <- [];
    draw_cards player 5 |> ignore;
    game.active_player_index <- (if game.active_player_index = (List.length game.players) - 1
                                 then 0 else game.active_player_index + 1);
    game.stage <- Action;
    game.stat.actions <- 1;
    game.stat.buys <- 1;
    game.stat.coins <- 0
  in
  match game.stage with
  | Action -> play_action ()
  | Treasure -> play_treasure ()
  | Buy -> play_buy ()
  | Cleanup -> play_cleanup ()

let play_til_end io game =
  let one () =
    play_one io game;
    if is_ended game then Some (end_stat game) else None
  in
  loop_til ~f:one
  |> List.iter ~f:(fun (name, vps, winner) ->
      sprintf "%s: %d %s" name vps (if winner then "(win)" else "") |> io.output)

let garden =
  let vps p =
    let cnt = [p.deck; p.hand; p.played; p.discard]
              |> List.map ~f:List.length |> List.fold ~init:0 ~f:(+) in
    cnt / 10
  in { title ="Garden"; cost = 4; feature = DynamicVictory vps }

let feast =
  let play io (game, self, trashed) =
    if not trashed then begin
      game.trash <- self :: game.trash;
      io.output ("Trashed " ^ self.title);
      [Fn.non pile_empty; (fun p -> p.sample.cost <= 5)]
      |> all_preds
      |> gain_card io game (active_player game);
      true
    end else false
  in { title = "Feast"; cost = 4; feature = SelfTrashAction play }

let smithy =
  let play _ game = draw_cards (active_player game) 3 |> ignore
  in { title = "Smithy"; cost = 4; feature = BasicAction play }

let remodel =
  let play io game =
    let player = active_player game in
    let trash_card () =
      match player.hand
            |> List.map ~f:(fun c -> (c.title, true))
            |> choose io MandatoryOne "Select a card to trash"
      with
      | Unselectable ->
        io.output "No card in hand to trash";
        None
      | Indexes indexes ->
        begin match split_by_indexes player.hand indexes with
        | ([card], hand) ->
          player.hand <- hand;
          game.trash <- card :: game.trash;
          "Trashed " ^ card.title |> io.output;
          Some card
        | (_, _) -> assert false
        end
      | Skip -> assert false
    in
    match trash_card () with
    | Some card ->
      [Fn.non pile_empty; (fun p -> p.sample.cost <= card.cost + 2)]
      |> all_preds
      |> gain_card io game player
    | None -> ()
  in { title = "Remodel"; cost = 4; feature = BasicAction play }

let throne_room =
  let play io game =
    let player = active_player game in
    match player.hand
          |> List.map ~f:(card_to_item is_action)
          |> choose io MandatoryOne "Select an action card to play twice"
    with
    | Unselectable ->
      io.output "No action card to play"
    | Indexes indexes ->
      begin match split_by_indexes player.hand indexes with
      | ([card], hand) ->
        player.hand <- hand;
        player.played <- card :: player.played;
        begin match card.feature with
        | BasicAction play ->
          "playing " ^ card.title ^ " first time" |> io.output;
          play io game;
          "playing " ^ card.title ^ " second time" |> io.output;
          play io game
        | SelfTrashAction play ->
          "playing " ^ card.title ^ " first time" |> io.output;
          let trashed = play io (game, card, false) in
          "playing " ^ card.title ^ " second time" |> io.output;
          play io (game, card, trashed) |> ignore
        | BasicTreasure _ | BasicVictory _ | DynamicVictory _ -> assert false
        end
      | (_, _) -> assert false
      end
    | Skip -> assert false
  in { title = "Throne Room"; cost = 4; feature = BasicAction play }

let create_console_io () = {
  input = read_line;
  output = print_endline;
}

let create_recorded_io inputs =
  let inputs = ref inputs in
  let outputs = ref [] in
  let input () = match !inputs with
    | [] -> failwith "no more recorded input to use"
    | hd :: tl -> inputs := tl; hd
  in
  let output message = outputs := message :: !outputs in
  let dump_outputs () = List.rev !outputs in
  ({ input; output }, dump_outputs)

let create_game names =
  let player_cnt = List.length names in
  { players = List.map names ~f:create_player;
    active_player_index = Random.int player_cnt;
    piles = create_start_piles [remodel; garden; feast; smithy; throne_room] player_cnt;
    trash = [];
    stage = Action;
    stat = { actions = 1; buys = 1; coins = 0 };
  }

let main () =
  let io = create_console_io () in
  let game = create_game ["wes";"bec"] in
  play_til_end io game
